expr/func_aspercent.go

```diff
@@ -0,0 +1,254 @@
+package expr
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/grafana/metrictank/api/models"
+	schema "gopkg.in/raintank/schema.v1"
+)
+
+type FuncAsPercent struct {
+	in          GraphiteFunc
+	totalFloat  float64
+	totalSeries GraphiteFunc
+	nodes       []expr
+}
+
+func NewAsPercent() GraphiteFunc {
+	return &FuncAsPercent{totalFloat: math.NaN()}
+}
+
+func (s *FuncAsPercent) Signature() ([]Arg, []Arg) {
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgIn{
+			key: "total",
+			opt: true,
+			args: []Arg{
+				ArgFloat{val: &s.totalFloat},
+				ArgSeriesList{val: &s.totalSeries},
+			},
+		},
+		ArgStringsOrInts{val: &s.nodes, opt: true, key: "nodes"},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncAsPercent) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncAsPercent) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var outSeries []models.Series
+	defer func() { cache[Req{}] = append(cache[Req{}], outSeries...) }()
+	var totals []models.Series
+	if s.totalSeries != nil {
+		totals, err = s.totalSeries.Exec(cache)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if s.nodes != nil {
+		if !math.IsNaN(s.totalFloat) {
+			return nil, errors.New("total must be None or a seriesList")
+		}
+		outSeries, err = s.execWithNodes(series, totals)
+	} else {
+		if totals != nil && len(totals) != 1 && len(totals) != len(series) {
+			return nil, errors.New("asPercent second argument (total) must be missing, a single digit, reference exactly 1 series or reference the same number of series as the first argument")
+		}
+		outSeries, err = s.execWithoutNodes(series, totals)
+	}
+	return outSeries, err
+}
+
+func (s *FuncAsPercent) execWithNodes(series, totals []models.Series) ([]models.Series, error) {
+	var outSeries []models.Series
+	// Set of keys
+	keys := make(map[string]struct{})
+	// Series grouped by key
+	metaSeries := groupSeriesByKey(series, s.nodes, &keys)
+	// The totals series for each key
+	var totalSeries map[string]models.Series
+
+	// calculate the sum
+	if math.IsNaN(s.totalFloat) && totals == nil {
+		totalSeries = getTotalSeries(metaSeries)
+		// calculate sum of totals series
+	} else if totals != nil {
+		totalSeriesLists := groupSeriesByKey(totals, s.nodes, &keys)
+		totalSeries = getTotalSeries(totalSeriesLists)
+	}
```

Comment:
* we compute totals even for keys we won't need (e.g. keys not in the input series)
* for each missing case we repeatedly get a slice and fill it with NaN's. we could reuse the same slice in this case. my suggestion would be declare a `var nones []schema.Point`. then whenever you need it, if it's nil, instantiate it. if it's not, just reuse it.

Reply:
I don't think that would play nicely with the pointSlicePool, would it?

Reply:
oh right, because at the end the same slice would get added to the pool multiple times, which is bad. so nevermind that then.

Reply:
On phone now
best would be to just add the slice to the cache thing once I think